File: diagonal-traverse/diagonal-traverse.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        
        if len(mat)==0:
            return []
        
        m=len(mat)
        n=len(mat[0])
        
        result=[1] *m *n
        
        idx=0
        i=0
        j=0
        dir=1
        
        while(idx<m*n):
            result[idx]=mat[i][j]
            
            if dir==1:
                if(j==n-1):
                    i+=1
                    dir=-1
                
                elif(i==0):
                    j+=1
                    dir=-1
                    
                else:
                    i-=1
                    j+=1
            
            else:
                if(i==m-1):
                    j+=1
                    dir=1
                
                elif(j==0):
                    i+=1
                    dir=1
                    
                else:
                    i+=1
                    j-=1
            
            idx+=1
        
        
        return result
        

File: diagonal-traverse/diagonal-traverse.c


/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* findDiagonalOrder(int** mat, int matSize, int* matColSize, int* returnSize){
    
    int m=matSize;
    int n=*matColSize;
    
    // int *res=NULL;
    int *res=(int*)malloc(sizeof(int)*m*n);
    
    int dir=1;
    int idx=0;
    int i=0;
    int j=0;
    
    *returnSize=m*n;
    
    // printf("%d",idx);
    
    if (!mat || !matSize || !matColSize || !*matColSize)
        return res;
    
    // res=malloc(sizeof(int)*m*n);
    
    
    while(idx<m*n){
        res[idx]=mat[i][j];
        if(dir==1){
            if(j==n-1){
                    i++;
                    dir=-1;
                    
            }else if(i==0){
                j++;
                    dir=-1;
            }else{
                    i--;
                    j++;
                }
                
            }else{
                if(i==m-1){
                    j++;
                    dir=1;
                }else if(j==0){
                    i++;
                    dir=1;
                }else{
                    i++;
                    j--;
                }
            }
        
        idx++;
        // printf(idx);
    }
    return res;

}